### ISSUE TITLE
AI turns take too long

### ISSUE THREAD

**Issue by jficz** [w84death/Tanks-of-Freedom#27]
Hi, it would be nice if the CPU turns were not as long as they are now. Sometimes it takes around half a minute even if all enemy units are hidden in fog of war and sometimes even longer when the battefield is visible. This is for Android. It's slightly better on a Linux box but it could still be faster.

Perhaps a settings gauge to set AI turn time speed?

-cM.

**Comment by jficz** [comment]
I believe that significant reason for slowing down is the "smooth" following camera - it takes a lot of time until it fixes over a location. It's unfeasible to play without a following camera on Android due to smaller screen but if turned off on a computer with big screen, the AI turn time is almost twice as fast.

**Comment by czlowiekimadlo** [comment]
Hi cptMikky, thank you for the feedback! I would have to check, but I think, that AI takes it's turns as fast as it can. Idea we have to somewhat help with this issue is to add a configurable camera speed, so it takes less time to show the action. I will add this to the planning board, and consider adding it to some future release.

**Comment by caryoscelus** [comment]
It would me much better if AI would not keep switching units as it does now. Since the game is deterministic anyway, AI should be able to compute all actions for given unit and perform them at once.

**Comment by czlowiekimadlo** [comment]
The game is already computing all moves for a unit, it switches, because it sees other moves, that are 'better'. I think we could add some weight to moves done by the unit, that moved last, so it would switch less, but I'm not sure if we want to go this way. 

Right now AI should balance the movement between many units, so it does not send single guys for raids against big enemy force just for them to die - it should try keeping it's forces closer together. To fix that we would have to add much more movement prediction, and that is not really an option due to performance issues with AI scripts. Maybe when Godot 2.2 adds C# support and we can make pathfinding faster we will be able to make AI smarter.

**Comment by caryoscelus** [comment]
Well, i guess it's hard to completely fix this, but i've seen some situations which "should be" obvious to AI, like tank & heli moving against infantry & tank respectively - and they did it in single moves, switching between each other (and there was enough ap to do both in the end, which they did).

Not knowing architecture i don't know how hard is it to fix such issue though.

**Comment by Shuunen** [comment]
I'm on android , playing at the campaign 8/14, AI turn take around a complete minute even with disabled camera follow, any advise ? 

Wait a minute between each round is really frustrating :p 

Aside from this, thanks for your game it rocks :+1:

**Comment by czlowiekimadlo** [comment]
I'm afraid, that if turning camera follow to off does not help, then there are no more workarounds at this time. 
It is possible, that I never suffered long AI turns, as I try to not let it gather too much AP - it will cause it to have more to do and take more time.

Increasing AI speed is quite low on our priority list right now, but we are keeping this issue open to get back to it in the future. Also, I'm afraid, that turn-based strategies just tend to be slow :)

**Comment by Shuunen** [comment]
Yes it was because I let AI to much AP and I had to stop the mission that was un-finnishable :)

**Comment by czlowiekimadlo** [comment]
That actually gave me an idea. We could add an AP cap for the AI, so it's turns time would be somewhat capped as well. I will discuss this idea with the rest of the team.
